signer/plugin.go

```diff
@@ -0,0 +1,259 @@
+	if err = json.Unmarshal(envContent.Payload.Content, &signedPayload); err != nil {
+		return nil, nil, fmt.Errorf("signed envelope payload can't be unmarshaled: %w", err)
+	}
+
+	// TODO: Verify plugin did not add any additional top level payload attributes. https://github.com/notaryproject/notation-go/issues/80
+	if !descriptorPartialEqual(desc, signedPayload.TargetArtifact) {
+		return nil, nil, errors.New("descriptor subject has changed")
+	}
+
+	return resp.SignatureEnvelope, &envContent.SignerInfo, nil
+}
+
+func (s *pluginSigner) mergeConfig(config map[string]string) map[string]string {
+	c := make(map[string]string, len(s.pluginConfig)+len(config))
+	// First clone s.PluginConfig.
+	for k, v := range s.pluginConfig {
+		c[k] = v
+	}
+	// Then set or override entries from config.
+	for k, v := range config {
+		c[k] = v
+	}
+	return c
+}
+
+func (s *pluginSigner) describeKey(ctx context.Context, config map[string]string) (*proto.DescribeKeyResponse, error) {
+	req := &proto.DescribeKeyRequest{
+		ContractVersion: proto.ContractVersion,
+		KeyID:           s.keyID,
+		PluginConfig:    config,
+	}
+	resp, err := s.plugin.DescribeKey(ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("describe-key command failed: %w", err)
+	}
+
+	return resp, nil
+}
+
+// descriptorPartialEqual checks if the both descriptors point to the same resource
+// and that newDesc hasn't replaced or overridden existing annotations.
+func descriptorPartialEqual(original, newDesc ocispec.Descriptor) bool {
+	if !content.Equal(original, newDesc) {
+		return false
+	}
+	// Plugins may append additional annotations but not replace/override existing.
+	for k, v := range original.Annotations {
+		if v2, ok := newDesc.Annotations[k]; !ok || v != v2 {
+			return false
+		}
+	}
+	return true
+}
+
+func parseCertChain(certChain [][]byte) ([]*x509.Certificate, error) {
+	certs := make([]*x509.Certificate, len(certChain))
+	for i, cert := range certChain {
+		cert, err := x509.ParseCertificate(cert)
+		if err != nil {
+			return nil, err
+		}
+		certs[i] = cert
+	}
+	return certs, nil
+}
+
+// remoteSigner implements signature.Signer
+type remoteSigner struct {
+	plugin       plugin.SignPlugin
+	pluginConfig map[string]string
+	keySpec      signature.KeySpec
+	keyID        string
+}
+
+// Sign signs the digest by calling the underlying plugin.
+func (s *remoteSigner) Sign(payload []byte) ([]byte, []*x509.Certificate, error) {
+	// Execute plugin sign command.
+	keySpec, err := proto.EncodeKeySpec(s.keySpec)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	keySpecHash, err := proto.HashAlgorithmFromKeySpec(s.keySpec)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	req := &proto.GenerateSignatureRequest{
+		ContractVersion: proto.ContractVersion,
+		KeyID:           s.keyID,
+		KeySpec:         keySpec,
+		Hash:            keySpecHash,
+		Payload:         payload,
+		PluginConfig:    s.pluginConfig,
+	}
+
+	resp, err := s.plugin.GenerateSignature(context.Background(), req)
```

Comment:
Can we add a `ctrx context.Context` field in the `remoteSigner` so that we can use the context instead of `context.Background()`.